crates/store/src/historical/mod.rs

```diff
@@ -0,0 +1,407 @@
+//! Historical tracking for `AccountTree` via mutation overlays
+
+use std::collections::VecDeque;
+use std::sync::{Arc, RwLock};
+
+use miden_crypto::EMPTY_WORD;
+use miden_crypto::merkle::{
+    LeafIndex,
+    MerkleError,
+    NodeIndex,
+    NodeMutation,
+    SMT_DEPTH,
+    SmtLeaf,
+    SparseMerklePath,
+};
+use miden_objects::account::AccountId;
+use miden_objects::block::{AccountMutationSet, AccountTree, AccountWitness, BlockNumber};
+use miden_objects::{AccountTreeError, Word};
+
+#[cfg(test)]
+mod tests;
+
+#[allow(missing_docs)]
+#[derive(thiserror::Error, Debug)]
+pub enum HistoricalError {
+    #[error(transparent)]
+    MerkleError(#[from] MerkleError),
+    #[error(transparent)]
+    AccountTreeError(#[from] AccountTreeError),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum HistoricalOffset {
+    Future,
+    ReversionsIdx(usize),
+    Latest,
+    TooAncient,
+}
+
+/// Captures reversion state for historical queries.
+#[derive(Debug, Clone)]
+struct HistoricalOverlay {
+    block_number: BlockNumber,
+    rev_set: AccountMutationSet,
```

Comment:
We could probably split `AccountMutationSet` into individual fields. It could look something like:
```rust
struct HistoricalOverlay {
    block_number: BlockNumber,
    root: Word,
    node_mutations: HashMap<NodeIndex, Word>,
    account_updates: HashMap<Word, Word>,
}
```

Using hash maps should be more efficient for lookups, but may take more time to construct.